src/Bolero/Components.fs

```diff
@@ -36,6 +36,12 @@ type Component() =
 
     let matchCache = Dictionary()
 
+    /// Compare the old model with the new to decide whether this component
+    /// needs to be re-rendered.
+    abstract ShouldRender : oldModel: 'model * newModel: 'model -> bool
+    default this.ShouldRender(oldModel, newModel) =
+        not <| obj.ReferenceEquals(oldModel, newModel)
+
```

Comment:
Good idea moving ShouldRender here! The argument-less override should come with it though :)

Reply:
Scratch this, based on your previous comment, indeed there is no Model here.

It feels weird to have a method taking models in a type that has no models though :/ I think it would make more sense to duplicate the method and have it both on `ElmishComponent` and `ProgramComponent`.

Reply:
Perhaps it can be better resolved via a naming convention shift away from model and towards 'state'?  It isn't strictly about doing a model/model comparison so much as an 'a -> 'a -> bool, and it seems reasonable that there will always be some type 'a that we may want to compare.  It's less of a mandatory workflow on the type so much as a helper method made available (and coincidentally consumed by) the child types.  

Along with that, this creates a possibly useful separation of type generic compared to comparison generic - both ElmishComponent and ProgramComponent have a 'model parameter, which is the same parameter used by ShouldRender.

By doing this, we end up freeing up a more generic comparison - a user implementation of ProgramComponent as-is would be locked to comparing its own 'model type - by shifting this, we enable it to do comparison on, for example, 'model * 'a

Reply:
Though, of course, that's also available by simply duplicating this and changing the name of the type parameter.  I suppose the sticking point for me is just that duplication :)

Reply:
Okay so looking at it more closely, there's a fundamental problem with putting `ShouldRender(oldModel, newModel)` on `Component`. The issue is that it makes the method itself generic, rather than the type it's on. The problem is that this method is meant to be implemented in a concrete way, for example like this:

```fsharp
type Model = { id: string; data: string }

// ...
override this.ShouldRender(oldModel, newModel) =
    // We know that we'll never change `id` so we just compare `data`.
    oldModel.data <> newModel.data
```

This is not possible if `ShouldRender` is generic. Instead it needs to be a concrete method on a generic type, which `Component` isn't. I think a good solution would be to introduce an intermediary `Component<'model>` type, which inherits from `Component`, is inherited by both `ProgramComponent<'model, 'msg>` and `ElmishComponent<'model, 'msg>`, and just has a virtual method `ShouldRender(oldModel, newModel)`.

Reply:
Oof, you're right.  And I think that shim type makes sense to me in this context.